### PR TITLE
Enable 'strict' Mypy linting

### PR DESCRIPTION
**Pull request by danieleades**
enables strict mypy linting

note that i've suppressed and allow-listed some existing errors.
This allow-list should not be allowed to grow over time, and ideally would be incrementally shrunk until it can be removed entirely.

The error suppressions can be resolved by refactoring, and by creating upstream type stubs for third-party libraries that are untyped in typeshed (out of scope of this PR!)

### CODE REVIEW

**Comment by danieleades** [comment]
"just" 49 errors to go!

**Comment by danieleades** [comment]
remaining errors:

```
copier/tools.py:192: error: Missing type parameters for generic type "Match"  [type-arg]
copier/template.py:385: error: Missing type parameters for generic type "dict"  [type-arg]
copier/main.py:866: error: Item "None" of "Template | None" has no attribute "url"  [union-attr]
copier/main.py:867: error: Item "None" of "Template | None" has no attribute "commit"  [union-attr]
copier/main.py:889: error: Argument 2 to "migration_tasks" of "Template" has incompatible type "Template | None"; expected "Template"  [arg-type]
copier/main.py:911: error: Argument "data" to "replace" of "Worker" has incompatible type "Mapping[str, Any]"; expected "dict[str, Any]"  [arg-type]
copier/main.py:914: error: Item "None" of "Template | None" has no attribute "url"  [union-attr]
copier/main.py:993: error: Argument 2 to "migration_tasks" of "Template" has incompatible type "Template | None"; expected "Template"  [arg-type]
copier/cli.py:83: error: Class cannot subclass "Application" (has type "Any")  [misc]
copier/cli.py:105: error: Class cannot subclass "Application" (has type "Any")  [misc]
copier/cli.py:158: error: Untyped decorator makes function "data_switch" untyped  [misc]
copier/cli.py:176: error: Untyped decorator makes function "data_file_switch" untyped  [misc]
tests/test_dirty_local.py:7: error: Module "pytest_gitconfig" does not explicitly export attribute "GitConfig"  [attr-defined]
tests/conftest.py:11: error: Module "pytest_gitconfig" does not explicitly export attribute "GitConfig"  [attr-defined]
Found 14 errors in 6 files (checked 46 source files)
```

**Comment by danieleades** [comment]
@sisp, @yajo this is now ready for review

File: pyproject.toml
Comment (pawamoy) [inline pyproject.toml:63]: I'd suggest avoiding upper bounds (and shortening to just major.minor, but that's just my own taste):

```suggestion
types-pygments = ">=2.17"
types-colorama = ">=0.4"
```

File: copier/cli.py
Comment (pawamoy) [inline copier/cli.py:88]: These formatting changes again :smile: 
I suppose we should review and merge #1557 first, to get rid of this once and for all.

Comment (danieleades) [reply copier/cli.py:88]: ha i'm really struggling with this dev environment. keeps picking up my system version of ruff and applying these changes!

Comment (danieleades) [reply copier/cli.py:88]: or i can clean them up manually when everything else is reviewed, i don't really mind

File: copier/main.py
Comment (pawamoy) [inline copier/main.py:795]: If `new_worker.run_copy()` returns something, that would be a breaking change, even if we didn't advertise the return value because of our `None` return annotation. Need to decide whether we're OK with the breaking change or if we should update the return annotation to reflect the actual return value.

File: copier/cli.py
Comment (pawamoy) [reply copier/cli.py:88]: I think you can also disable auto-formatting in your editor settings while working on this (be careful not to include changes to VSCode's settings.json in the PR though). Unless you mean that the precommit tasks use your system Ruff instead of the Copier environment one? In any case, don't bother too much, this is not fun and I can revert it myself before we merge :slightly_smiling_face:

File: copier/main.py
Comment (danieleades) [reply copier/main.py:795]: i'm not sure what you mean?
`new_worker.run_copy()` doesn't return anything, and there's no return statement in that function anyway

File: tests/test_dirty_local.py
Comment (pawamoy) [inline tests/test_dirty_local.py:7]: Did mypy complain? The [docs](https://pypi.org/project/pytest-gitconfig/) show that we should import from `pytest_gitconfig` directly, and the [code](https://github.com/noirbizarre/pytest-gitconfig/blob/main/src/pytest_gitconfig/__init__.py) correctly imports `GitConfig` in the top-level module.

By the way the docs show that this import should be guarded under `if TYPE_CHECKING`, does it apply to us here as well? Should we `from __future__ import annotations` in all tests module (that's what I do in my own projects)?

File: copier/main.py
Comment (pawamoy) [reply copier/main.py:795]: Great, so the condition in my comment is never true, and we can discard the whole comment :slightly_smiling_face:

File: tests/test_dirty_local.py
Comment (danieleades) [reply tests/test_dirty_local.py:7]: > Did mypy complain? The [docs](https://pypi.org/project/pytest-gitconfig/) show that we should import from `pytest_gitconfig` directly, and the [code](https://github.com/noirbizarre/pytest-gitconfig/blob/main/src/pytest_gitconfig/__init__.py) correctly imports `GitConfig` in the top-level module.

mypy complains because `GitConfig` is not explicitly re-exported from `pytest_gitconfig`. That could be fixed with an upstream change


> By the way the docs show that this import should be guarded under `if TYPE_CHECKING`, does it apply to us here as well?

not required, but something i could add in a future PR

> Should we `from __future__ import annotations` in all tests module (that's what I do in my own projects)?

also not required, since ruff can detect when it is needed and automatically add it

**Comment by danieleades** [comment]
@pawamoy i could probably just keep tinkering with this for ever, but i think this is a place where I can draw a line under it. This is now ready for review

**Comment by danieleades** [comment]
> @sisp, @yajo this is now ready for review

ping!

**Review by pawamoy — COMMENTED** [review]
IMO it looks good, except for the style changes in cli.py. I'll revert them and push a fixup.

File: copier/main.py
Comment (sisp) [inline copier/main.py:2]: Those empty lines were introduced by Ruff v0.3.0 (IIRC). Would you mind reverting those (and the other occurences), too? They aren't related to this PR.

File: pyproject.toml
Comment (sisp) [inline pyproject.toml:149]: Would you mind moving this line up, so the modules are sorted?

File: pyproject.toml
Comment (sisp) [inline pyproject.toml:63]: Would you mind sorting the dependencies? I just noticed that `types-pyyaml` was already in the "wrong" place, too. I'd say, just sort them all now.

File: poetry.lock
Comment (sisp) [inline poetry.lock:1]: We've had a similar issue before: Some entries (for `pyyaml` and `pexpect`) get deleted for no obvious reason, and I don't think they should. If @danieleades can't fix it, would you mind fixing it again, @pawamoy? (I don't have a laptop right now.)

**Review by sisp — CHANGES_REQUESTED** [review]
Fantastic work, @danieleades! 👌 Just a few minor remarks from my side.

File: poetry.lock
Comment (pawamoy) [reply poetry.lock:1]: Urrh, my bad, didn't think about expanding poetry.lock's diff. One of the inconvenient of GitHub collapsing it by default, even for small diffs... Will do.

Comment (pawamoy) [reply poetry.lock:1]: Done.

**Comment by danieleades** [comment]
what's causing those test failures?

**Review by sisp — CHANGES_REQUESTED** [review]
The style changes in `cli.py` are still present (cc @pawamoy) and also the introduced empty line between a module doctstring and the first import, which is not needed by the Ruff version locked via Nix (IIRC).

**Comment by pawamoy** [comment]
Reverted style changes in cli.py. Will now revert blank lines at the top of modules. Race to the finish!

**Comment by pawamoy** [comment]
Done.

**Comment by sisp** [comment]
> what's causing those test failures?

I _suspect_ that postponed annotations don't play well with Plumbum in `cli.py`. Could you try reverting back to classical type annotations and see whether tests pass again?

**Comment by danieleades** [comment]
> > what's causing those test failures?
> 
> I _suspect_ that postponed annotations don't play well with Plumbum in `cli.py`. Could you try reverting back to classical type annotations and see whether tests pass again?

strange. i've not made any changes since these tests were last passing

File: copier/cli.py
Comment (sisp) [inline copier/cli.py:168]: Should this really be an `Iterable` and not a `Sequence`?

File: copier/cli.py
Comment (danieleades) [inline copier/cli.py:111]: ah i think this was the issue, not sure when this got reverted

File: copier/cli.py
Comment (pawamoy) [reply copier/cli.py:168]: It could be a sequence, but it doesn't *need* to, we only iterate on it and do not access its element by index.

**Review by pawamoy — APPROVED** [review]
LGTM

File: copier/cli.py
Comment (pawamoy) [reply copier/cli.py:168]: @danieleades I think it's better if you let the person asking the question mark a comment as resolved :slightly_smiling_face:

**Review by sisp — APPROVED** [review]
LGTM, too! :tada: Thanks for the fantastic work, @danieleades and @pawamoy! :pray:

**Comment by sisp** [comment]
Resolves #1441.

**Pull request merged**
